lib/src/authentication/token_store.dart

```diff
@@ -0,0 +1,156 @@
+              'tokens.json format is invalid or not supported.');
+        }
+
+        result.addAll((json['hosted'] as List)
+            .cast<Map<String, dynamic>>()
+            .map((it) => Token.fromJson(it)));
+      }
+    } on FormatException catch (error, stackTrace) {
+      log.error('Failed to load tokens.json', error, stackTrace);
+
+      // When an invalid, damaged or not compatible version of token.json is
+      // found, we remove it after showing error message. Otherwise the error
+      // message will be displayed on each pub command.
+      // Or instead we could write instructrions to execute
+      //`pub token remove --all` if user couldn't solve the issue.
+      deleteEntry(_tokensFile);
+    }
+
+    return result;
+  }
+
+  /// Writes [tokens] into "tokens.json".
+  void _saveTokens(List<Token> tokens) {
+    writeTextFile(
+        _tokensFile,
+        jsonEncode(<String, dynamic>{
+          'version': 1,
+          'hosted': tokens.map((it) => it.toJson()).toList(),
+        }));
+  }
+
+  /// Writes latest state of the store to disk.
+  void flush() {
+    if (_tokens == null) {
+      throw Exception('Schemes should be loaded before saving.');
+    }
+    _saveTokens(_tokens!);
+  }
+
+  /// Adds [token] into store and writes into disk.
+  void addToken(Token token) {
+    // Remove duplicate tokens
+    tokens.removeWhere((it) => it.url == token.url);
+    tokens.add(token);
+    flush();
+  }
+
+  /// Removes tokens with matching [url] from store. Returns whether or not
+  /// there's a stored token with matching url.
+  bool removeMatchingTokens(String url) {
+    final normalizedUrl = validateAndNormalizeHostedUrl(url);
+
+    var i = 0;
+    var found = false;
+    while (i < tokens.length) {
+      if (tokens[i].url == normalizedUrl) {
+        tokens.removeAt(i);
+        found = true;
+      } else {
+        i++;
+      }
+    }
+
+    flush();
+
+    return found;
+  }
+
+  /// Returns [Token] for authenticating given url or null if no matching token
+  /// is found.
+  Token? findToken(Uri url) {
+    Token? matchedToken;
+    for (final token in tokens) {
+      if (token.url == url) {
+        if (matchedToken == null) {
+          matchedToken = token;
+        } else {
+          log.warning(
+            'Found multiple matching authentication tokens for "$url". '
+            'First matching token will be used for authentication.',
+          );
+          break;
+        }
+      }
+    }
+
+    return matchedToken;
+  }
+
+  /// Returns whether or not store contains a token that could be used for
+  /// authenticating given [url].
+  bool hasToken(Uri url) {
+    return tokens.any((it) => it.url == url);
+  }
+
+  /// Deletes tokens.json file from the disk.
+  void deleteTokensFile() {
+    deleteEntry(_tokensFile);
+    log.message('tokens.json is deleted.');
+  }
+
+  /// Full path to the "tokens.json" file.
+  String get _tokensFile => path.join(cacheRootDir, 'tokens.json');
```

Comment:
We should not store these in the pub cache.

When (hopefully) https://github.com/dart-lang/pub/pull/3092 has landed we can use `dartConfigDir/tokens.json`.

Reply:
Yes I know, we've decided to do this in a separate PR because credentials.json (for pub.dartlang.org) is also stored there.

Reply:
Yes, let's do this separately.